### PR TITLE
add support for failover to RADIUS client

### PR DESCRIPTION
**Pull request by 0xAX**
This commit adds support for fail-over. Set of secondary RADIUS servers could be passed to the RADIUS client API eradius_client:send_request/3 via options or to RADIUS proxy via configuration.

If the response wasn't received after a number of requests specified by `retries` configuration option for the given RADIUS server - such RADIUS servers will be marked as non-active and RADIUS requests will not be sent for such `non-active` RADIUS servers. The new configuration option is represented as well in this PR - `eradius.unreachable_timeout`. This option specifies how long such `non-active` RADIUS servers should be considered as unreachable. Such `non-active` RADIUS server will be marked as active and returned to the servers pool after the timeout specified as a value of the `eradius.unreachable_timeout` option will be expired.

Secondary RADIUS servers could be specified via RADIUS proxy configuration, with the new configuration option - pool name.

For example:

```
  [{default_route, {{127, 0, 0, 1}, 1812, <<"secret">>}, pool_name},
```

Where the `pool_name` atom specifies name of a pool of secondary RADIUS servers. All pools are configured via:

```
   {servers_pool, [{pool_name, [
                     {{127, 0, 0, 2}, 1812, <<"secret">>, [{retries, 3}]},
                     {{127, 0, 0, 3}, 1812, <<"secret">>}]}]}
```

In a case when RADIUS proxy (eradius_proxy handler) is not used, a list of RADIUS upstream servers could be passed to the `eradius_client:send_radius_request/3` via options, for example:

```
eradius_client:send_request(Server, Request, [{failover, [{"localhost", 1814, <<"secret">>}]}])
```

All such active RADIUS servers will be stored in the ets table and in a case of failed RADIUS request - eradius_client will fetch next one active RADIUS server from the ets table and it will be used for next tries of sending failed RADIUS request. If `failover` option was not passed to the client through the options or RADIUS proxy configuration there should not be any performance impact as RADIUS client will try to a RADIUS request to only one RADIUS server that is defined in eradius_client:send_request/3 options.

### CODE REVIEW

**Comment by vkatsuba** [comment]
Add documentation section in scope of https://github.com/travelping/eradius/issues/165.

File: src/eradius_proxy.erl [hunk 1]
```diff
@@ -39,25 +50,25 @@
                           {timeout, ?DEFAULT_TIMEOUT},
                           {retries, ?DEFAULT_RETRIES}]).
 
--type route() :: eradius_client:nas_address().
--type routes() :: [{Name :: string(), route()}].
+-type route() :: {default_route, eradius_client:nas_address()}.
+-type routes() :: {routes, [{Name :: string(), route()}]}.
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:54]:
```erlang
-type routes() :: {routes, [{Name :: string(), route()}]} | {routes, [{Name :: string(), route(), PoolName :: atom()}]}.
```

File: src/eradius_proxy.erl [hunk 2]
```diff
@@ -39,25 +50,25 @@
                           {timeout, ?DEFAULT_TIMEOUT},
                           {retries, ?DEFAULT_RETRIES}]).
 
--type route() :: eradius_client:nas_address().
--type routes() :: [{Name :: string(), route()}].
+-type route() :: {default_route, eradius_client:nas_address()}.
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:53]: Looks like need update `eradius_client:nas_address()`
```erlang
-type nas_address() :: {string() | binary() | inet:ip_address(),  eradius_server:port_number(),  eradius_lib:secret()}
                     | {string() | binary() | inet:ip_address(),  eradius_server:port_number(),  eradius_lib:secret(), PoolName :: atom()}.
```

File: src/eradius_proxy.erl [hunk 3]
```diff
@@ -66,13 +77,21 @@ validate_arguments(Args) ->
             {true, [{default_route, DefaultRoute}, {options, Options}, {routes, NewRoutes}]}
     end.
 
-compile_routes(Routes) ->
-    RoutesOpts = lists:map(fun ({Name, Relay}) ->
+compile_routes(false) ->
+    [];
+compile_routes({routes, Routes}) ->
+    RoutesOpts = lists:map(fun (Route) ->
+                          {Name, Relay, Pool} = case Route of
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:84]: looks like you need align left `{Name, Relay, Pool}` on a level with `case re:compile(Name) of`

File: src/eradius_proxy.erl [hunk 4]
```diff
@@ -211,14 +261,10 @@ strip(Username, prefix, true, Separator) ->
...
-%    case DefaultRouter of
-%	undefined ->
-%	    MetricName = eradius_metrics:get_metric_name(IP, Port, routes_not_resolved, proxy),
-%	    exometer:update_or_create(MetricName, 1, spiral, [{time_span, 1000}]);
-%	_ ->
-%	    MetricName = eradius_metrics:get_metric_name(IP, Port, routes_resolved, proxy),
-%	    exometer:update_or_create(MetricName, 1, spiral, [{time_span, 1000}])
-%    end.
-
+get_proxy_opt(OptName, Args) ->
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:264]: A little easier to read:
```erlang
get_proxy_opt(_, [])                            -> undefined;
get_proxy_opt(OptName, [{OptName, OptValue}|_]) -> OptValue;
get_proxy_opt(OptName, [_|T])                   -> get_proxy_opt(OptName, T).
```

**Review by vkatsuba — APPROVED** [review]

File: src/eradius_proxy.erl [hunk 5]
```diff
@@ -39,25 +50,25 @@
...
-    Routes = proplists:get_value(routes, Args, []),
     {NewUsername, Route} = resolve_routes(Username, DefaultRoute, Routes, Options),
     Retries = proplists:get_value(retries, Options, ?DEFAULT_RETRIES),
     Timeout = proplists:get_value(timeout, Options, ?DEFAULT_TIMEOUT),
     SendOpts = [{retries, Retries}, {timeout, Timeout}],
     send_to_server(new_request(Request, Username, NewUsername), Route, SendOpts).
 
 validate_arguments(Args) ->
-    DefaultRoute = proplists:get_value(default_route, Args, {undefined, 0, []}),
+    DefaultRoute = lists:keyfind(default_route, 1, Args),
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:69]: `lists:keyfind/3` will return `false` - this is expected as default value for `DefaultRoute` for `validate_route/1`?

File: src/eradius_proxy.erl [hunk 6]
```diff
@@ -66,13 +77,21 @@ validate_arguments(Args) ->
             {true, [{default_route, DefaultRoute}, {options, Options}, {routes, NewRoutes}]}
     end.
 
-compile_routes(Routes) ->
-    RoutesOpts = lists:map(fun ({Name, Relay}) ->
+compile_routes(false) ->
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:80]: `undefined` more better for perception as `false`

File: src/eradius_proxy.erl [hunk 7]
```diff
@@ -158,7 +190,13 @@ new_request(Request, _Username, NewUsername) ->
                      Routes :: routes(), Options :: [proplists:property()]) ->
                      {NewUsername :: string(), Route :: route()}.
 resolve_routes( undefined, DefaultRoute, _Routes, _Options) ->
-    {undefined, DefaultRoute};
+    Route = case DefaultRoute of
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:193]: A little easier to read:

```erlang
resolve_routes(undefined, {_, {Server, Port, Secret}, Pool}, _, _) -> 
  {undefined, {Server, Port, Secret, Pool}};
resolve_routes(undefined, {_, {_, _, _} = Route}, _, _) -> 
  {undefined, Route}.
```

File: src/eradius_proxy.erl [hunk 8]
```diff
@@ -39,25 +50,25 @@
...
     SendOpts = [{retries, Retries}, {timeout, Timeout}],
     send_to_server(new_request(Request, Username, NewUsername), Route, SendOpts).
 
 validate_arguments(Args) ->
-    DefaultRoute = proplists:get_value(default_route, Args, {undefined, 0, []}),
+    DefaultRoute = lists:keyfind(default_route, 1, Args),
     Options = proplists:get_value(options, Args, ?DEFAULT_OPTIONS),
-    Routes = proplists:get_value(routes, Args, []),
+    Routes = lists:keyfind(routes, 1, Args),
     case {validate_route(DefaultRoute), validate_options(Options), compile_routes(Routes)} of
```
Comment (ebengt) [inline src/eradius_proxy.erl:72]: Is this case finding correct/incorrect default_route, options, route? It would be clearer if the return values where
incorrect_default_route, incorrect_options, incorrect_routes.
When correct only the list need to be returned. It is different enough from the incorrect.

**Review by ebengt — APPROVED** [review]
In the description (thank you for having one. Much appreciated) it says

If the response wasn't received after a number of requests specified by 'retries' RADIUS client options - such RADIUS servers will be marked as non-active and will not be used for next requests, while configurable timeout (eradius.unreachable_timeout) is not expired.

I do not understand if the timeout is for the "will not be used" or for something else. Otherwise good explanation. Is it included in the code/README/xxx somewhere?

**Comment by 0xAX** [comment]
@ebengt 

> I do not understand if the timeout is for the "will not be used" or for something else. 

yes, I meant that requests will not be sent to such RADIUS servers, but yes, I will rephrase this to be more clear

> Is it included in the code/README/xxx somewhere?

For this moment there is no separate docs for this, besides header of eradius_proxy.erl. But sure, source code is not the best place for documentation. As mentioned before we will prepare eradius documentation with @vkatsuba  (including these changes when it will be merged) in separate pull request as the issue already created for this.

File: src/eradius_client.erl [hunk 9]
```diff
@@ -313,14 +358,55 @@ code_change(_OldVsn, State, _Extra) -> {ok, State}.
...
+                            case Server of
+                                {Addr, Port, _Secret, Opts} ->
+                                    Retries = proplists:get_value(retries, Opts, ?DEFAULT_RETRIES),
+                                    store_radius_server_from_pool(Addr, Port, Retries);
+                                {Addr, Port, _Secret} ->
+                                    store_radius_server_from_pool(Addr, Port, ?DEFAULT_RETRIES)
+                            end
+                        end, Servers)
+                    end, Pool)
+            end,
```
Comment (RoadRunnr) [inline src/eradius_client.erl:380]: do many levels for one function, break this down to a sensible helper

File: src/eradius_proxy.erl [hunk 10]
```diff
@@ -169,7 +207,14 @@ resolve_routes(Username, DefaultRoute, Routes, Options) ->
...
                 {Key, {_IP, _Port, _Secret} = Route} -> {NewUsername, Route};
-                _ -> {NewUsername, DefaultRoute}
+                {Key, {IP, Port, Secret}, Pool} -> {NewUsername, {IP, Port, Secret, Pool}};
+                _ ->
+                    case DefaultRoute of
+                        {default_route, {Server, Port, Secret}, Pool} ->
+                            {NewUsername, {Server, Port, Secret, Pool}};
+                        {default_route, {Server, Port, Secret}} ->
+                            {NewUsername, {Server, Port, Secret}}
+                    end
```
Comment (RoadRunnr) [inline src/eradius_proxy.erl:217]: case levels too deep, use a helper.

File: src/eradius_proxy.erl (hunk 1 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: fixed

File: src/eradius_proxy.erl (hunk 2 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: pool name is not related to nas_address()

File: src/eradius_proxy.erl (hunk 4 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: thanks, done

File: src/eradius_proxy.erl (hunk 6 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: this is fixed as well

File: src/eradius_client.erl (hunk 9 above)
Comment (0xAX) [reply src/eradius_client.erl]: @RoadRunnr yes, you're right. I've reworked this

File: src/eradius_proxy.erl (hunk 10 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: done

File: src/eradius_proxy.erl (hunk 8 above)
Comment (0xAX) [reply src/eradius_proxy.erl:74]: @ebengt 

> Is this case finding correct/incorrect default_route, options, route

yes, it is this. Now in a case of incorect default route, routes or options we already return atom `default_route`, `options` or `routes` and this atom will be used in error logs here - https://github.com/travelping/eradius/blob/master/src/eradius_config.erl#L79, so it should be pretty clear in a case of such errors.

File: src/eradius_proxy.erl (hunk 3 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: fixed

File: src/eradius_proxy.erl (hunk 5 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: this is reworked as well

File: src/eradius_proxy.erl (hunk 7 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: done

File: src/eradius_client.erl [hunk 11]
```diff
@@ -125,42 +126,87 @@ send_remote_request(Node, {IP, Port, Secret}, Request, Options) when ?GOOD_CMD(R
...
-proceed_response(Request, Response, _Peer, TS1, MetricsInfo) ->
+
+proceed_response(Request, Response, {_ServerName, {ServerIP, Port}}, TS1, MetricsInfo, Options) ->
     update_client_responses(MetricsInfo),
     update_client_request(Request#radius_request.cmd, MetricsInfo, eradius_lib:timestamp(milli_seconds) - TS1, Request),
-    Response.
+    maybe_failover(Request, Response, {ServerIP, Port}, Options).
+
+maybe_failover(Request, Response, {ServerIP, Port}, Options) ->
+    case proplists:get_value(failover, Options) of
```
Comment (vkatsuba) [inline src/eradius_client.erl:171]:
```erlang
    case proplists:get_value(failover, Options, []) of
        [] ->
            Response;
        [{NewServerIP, NewPort, NewSecret, _ServerRetries} | UpstreamServers] ->
            handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options);
        [{NewServerIP, NewPort, NewSecret} | UpstreamServers] ->
            handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options)
    end.
```

File: src/eradius_proxy.erl [hunk 12]
```diff
@@ -211,14 +245,10 @@ strip(Username, prefix, true, Separator) ->
...
-%	    MetricName = eradius_metrics:get_metric_name(IP, Port, routes_not_resolved, proxy),
-%	    exometer:update_or_create(MetricName, 1, spiral, [{time_span, 1000}]);
-%	_ ->
-%	    MetricName = eradius_metrics:get_metric_name(IP, Port, routes_resolved, proxy),
-%	    exometer:update_or_create(MetricName, 1, spiral, [{time_span, 1000}])
-%    end.
+route({RouteName, RouteRelay}) -> {RouteName, RouteRelay, undefined};
+route({RouteName, RouteRelay, Pool}) -> {RouteName, RouteRelay, Pool}.
 
+get_proxy_opt(_, [], Default) -> Default;
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:251]: Aligning to right `-> Default;` at the level with the code below

**Comment by vkatsuba** [comment]
Not sure, but looks like need add test case with invalid config of poll name.

File: src/eradius_proxy.erl [hunk 13]
```diff
@@ -211,14 +245,10 @@ strip(Username, prefix, true, Separator) ->
...
-%    case DefaultRouter of
-%	undefined ->
-%	    MetricName = eradius_metrics:get_metric_name(IP, Port, routes_not_resolved, proxy),
-%	    exometer:update_or_create(MetricName, 1, spiral, [{time_span, 1000}]);
-%	_ ->
-%	    MetricName = eradius_metrics:get_metric_name(IP, Port, routes_resolved, proxy),
-%	    exometer:update_or_create(MetricName, 1, spiral, [{time_span, 1000}])
-%    end.
+route({RouteName, RouteRelay}) -> {RouteName, RouteRelay, undefined};
+route({RouteName, RouteRelay, Pool}) -> {RouteName, RouteRelay, Pool}.
```
Comment (vkatsuba) [inline src/eradius_proxy.erl:249]: `route(Data) -> Data.`
or
`route({_RouteName, _RouteRelay, _Pool} = Data) -> Data.`

File: src/eradius_client.erl (hunk 11 above)
Comment (0xAX) [reply src/eradius_client.erl]: thanks, fixed

File: src/eradius_proxy.erl (hunk 13 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: done

**Comment by 0xAX** [comment]
> Not sure, but looks like need add test case with invalid config of poll name.
added tests for such cases

File: src/eradius_proxy.erl (hunk 12 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: fixed

**Review by vkatsuba — APPROVED** [review]

**Comment by ebengt** [comment]
I am still struggling with the Merge Request description.
eradius.unreachable_timeout, is it for the response or the request part of the sentence? Or both?

**Comment by 0xAX** [comment]
> I am still struggling with the Merge Request description.
> eradius.unreachable_timeout, is it for the response or the request part of the sentence? Or both?

@ebengt updated PR description

File: src/eradius_proxy.erl [hunk 14]
```diff
@@ -111,8 +136,15 @@ decode_request(Result, ReqID, Secret, Auth) ->
     end.
 
 % @private
--spec validate_route({Host :: list() | binary() | inet:ip_address(), Port :: eradius_server:port_number(), Secret :: eradius_lib:secret()}) ->
-    boolean().
+-spec validate_route({Host :: list() | binary() | inet:ip_address(),
```
Comment (ebengt) [inline src/eradius_proxy.erl:139]: Maybe have -type() for these two arguments?

**Review by ebengt — APPROVED** [review]

File: src/eradius_client.erl [hunk 15]
```diff
@@ -125,42 +126,86 @@ send_remote_request(Node, {IP, Port, Secret}, Request, Options) when ?GOOD_CMD(R
...
+        [{NewServerIP, NewPort, NewSecret, _ServerRetries} | UpstreamServers] ->
+            handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options);
+        [{NewServerIP, NewPort, NewSecret} | UpstreamServers] ->
+            handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options)
+    end.
+
+handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options) ->
+    case ets:lookup(?MODULE, {ServerIP, Port}) of
+        [{{ServerIP, Port}, Retries, InitialRetries}] ->
+            RetriesDone = proplists:get_value(retries, Options, ?DEFAULT_RETRIES),
```
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl:183]: Yes, this is amount of failed retries, will rename it

File: src/eradius_client.erl [hunk 16]
```diff
@@ -323,6 +378,31 @@ configure(State) ->
...
+    prepare_pool(Servers).
+
+%% private
+store_radius_server_from_pool(Addr, Port, Retries) when is_tuple(Addr) ->
+    ets:insert(?MODULE, {{Addr, Port}, Retries, Retries});
+store_radius_server_from_pool(Addr, Port, Retries) ->
+    case inet:gethostbyname(Addr) of
+        {ok, #hostent{h_addrtype = inet, h_addr_list = IPs}} ->
+            [ets:insert(?MODULE, {{IP, Port}, Retries, Retries}) || IP <- IPs];
+        _ ->
```
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl:401]: Here erlang logger is used and it depends on its configuration where logs will be stored. So, if an application that uses eradius will configure file backend for logger, this should be stored in file. In addition this should be called at eradius startup in general so it should fail very fast

File: src/eradius_proxy.erl (hunk 14 above)
Comment (0xAX) [reply src/eradius_proxy.erl:139]: Yes, but it looks like it could be just replaced with `route` type here

File: src/eradius_proxy.erl [hunk 17]
```diff
@@ -39,25 +50,27 @@
                           {timeout, ?DEFAULT_TIMEOUT},
                           {retries, ?DEFAULT_RETRIES}]).
 
--type route() :: eradius_client:nas_address().
--type routes() :: [{Name :: string(), route()}].
+-type route() :: {eradius_client:nas_address()} |
+                 {eradius_client:nas_address(), PoolName :: atom()}.
+-type routes() :: {[{Name :: string(), eradius_client:nas_address()}]} |
```
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_proxy.erl:55]: No, they are not:

  * `route` specifies `{nas_address, pool}`
  * `routes` specifies list of `{name, nas_address, pool}`

but yes, this declaration should be fixed as `route` and `routes` should not be tuples here

File: src/eradius_client.erl [hunk 18]
```diff
@@ -323,6 +378,31 @@ configure(State) ->
             {error, {bad_client_ip, ClientIP}}
     end.
 
+%% private
+prepare_pools() ->
+    ets:new(?MODULE, [ordered_set, public, named_table, {keypos, 1}, {write_concurrency,true}]),
+    lists:foreach(fun({_PoolName, Servers}) -> prepare_pool(Servers) end, application:get_env(eradius, servers_pool, [])).
+prepare_pool([]) -> ok;
```
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl:385]: No, functionality wasn't changed. It is just about simplification to split a couple of big nested list:foreach/2 calls for small help functions

File: src/eradius_client.erl [hunk 19]
```diff
@@ -313,6 +358,16 @@ code_change(_OldVsn, State, _Extra) -> {ok, State}.
...
 configure(State) ->
+    case ets:info(?MODULE) of
+        undefined ->
+            prepare_pools();
+        _ ->
+            % if ets table is already exists - which could be in a case of
+            % reconfigure, just re-create the table and fill it with newly
+            % configured pools of RADIUS upstream servers
+            ets:delete(?MODULE),
+            configure(State)
```
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl:369]: yes, it is probably will be more cleaner to call perpare_pools/0 here instead of recursive call of `configure`, will fix this

File: src/eradius_client.erl [hunk 20]
```diff
@@ -125,42 +126,86 @@ send_remote_request(Node, {IP, Port, Secret}, Request, Options) when ?GOOD_CMD(R
...
+    case ets:lookup(?MODULE, {ServerIP, Port}) of
+        [{{ServerIP, Port}, Retries, InitialRetries}] ->
+            RetriesDone = proplists:get_value(retries, Options, ?DEFAULT_RETRIES),
+            if (Retries - RetriesDone) =< 0 ->
+                    ets:delete(?MODULE, {ServerIP, Port}),
+                    Timeout = application:get_env(eradius, unreachable_timeout, 60),
+                    timer:apply_after(Timeout * 1000, ?MODULE, restore_upstream_server,
+                                      [{ServerIP, Port, InitialRetries, InitialRetries}]);
+               true ->
+                    ets:update_counter(?MODULE, {ServerIP, Port}, -RetriesDone)
```
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl:190]: `retries` option could be passed to `eradius_client:send_radius_request/3`, so client will try to send `N` RADIUS requests to the given server where `N` is value of the `retries`. So it could be less than `1`, that is why `-RetriesDone` here instead of `-1`. But again, yes, the name could be not very clear here so I will rename it to `FailedTries` or so and will put the comment.

File: src/eradius_client.erl (hunk 16 above)
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl:401]: And if you're about `_` error itself, yes will add it to the log

File: src/eradius_proxy.erl (hunk 14 above)
Comment (0xAX) [reply src/eradius_proxy.erl]: this is fixed now

File: src/eradius_proxy.erl (hunk 17 above)
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_proxy.erl]: these types are fixed as well

File: src/eradius_client.erl (hunk 16 above)
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl]: Added error to the log

File: src/eradius_client.erl (hunk 19 above)
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl]: done

File: src/eradius_client.erl (hunk 20 above)
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl]: added comment

File: src/eradius_client.erl [hunk 21]
```diff
@@ -125,42 +126,86 @@ send_remote_request(Node, {IP, Port, Secret}, Request, Options) when ?GOOD_CMD(R
...
+    case proplists:get_value(failover, Options, []) of
+        [] ->
+            Response;
+        [{NewServerIP, NewPort, NewSecret, _ServerRetries} | UpstreamServers] ->
+            handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options);
+        [{NewServerIP, NewPort, NewSecret} | UpstreamServers] ->
+            handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options)
+    end.
+
+handle_failed_request(Request, {ServerIP, Port}, {NewServerIP, NewPort, NewSecret}, UpstreamServers, Response, Options) ->
```
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl]: Yes, you're right, there is no need in it and removing of it will allow to simplify this function

File: src/eradius_client.erl (hunk 15 above)
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl]: renamed it to `FailedTries`

File: src/eradius_client.erl [hunk 22]
```diff
@@ -125,42 +126,86 @@ send_remote_request(Node, {IP, Port, Secret}, Request, Options) when ?GOOD_CMD(R
...
+    case ets:lookup(?MODULE, {ServerIP, Port}) of
+        [{{ServerIP, Port}, Retries, InitialRetries}] ->
+            RetriesDone = proplists:get_value(retries, Options, ?DEFAULT_RETRIES),
+            if (Retries - RetriesDone) =< 0 ->
```
[the remark this answers is not in the document]
Comment (0xAX) [reply src/eradius_client.erl]: reworked this

**Review by ebengt — APPROVED** [review]
Looks good.

**Review by vkatsuba — APPROVED** [review]

File: src/eradius_client.erl [hunk 23]
```diff
@@ -125,42 +126,90 @@ send_remote_request(Node, {IP, Port, Secret}, Request, Options) when ?GOOD_CMD(R
...
+
+proceed_response(Request, {ok, Response, Secret, Authenticator}, _Peer = {_ServerName, {ServerIP, Port}}, TS1, MetricsInfo, Options) ->
     update_client_request(Request#radius_request.cmd, MetricsInfo, eradius_lib:timestamp(milli_seconds) - TS1, Request),
     update_client_responses(MetricsInfo),
     case eradius_lib:decode_request(Response, Secret, Authenticator) of
         {bad_pdu, "Message-Authenticator Attribute is invalid" = Reason} ->
             update_client_response(bad_authenticator, MetricsInfo, Request),
             ?LOG(error, "~s INF: Noreply for request ~p. Could not decode the request, reason: ~s", [printable_peer(ServerIP, Port), Request, Reason]),
-            noreply;
+            maybe_failover(Request, noreply, {ServerIP, Port}, Options);
```
Comment (RoadRunnr) [inline src/eradius_client.erl:147]: this is IMHO not a fail-over reason, a misconfigured shared secret is a local problem and should cause an immediate abort of the request.

even if we consider a fail-over for this problem, it should be an immediate action. As I understand the logic, it would only fail over after exhausting the retries counter.

File: src/eradius_client.erl [hunk 24]
```diff
@@ -125,42 +126,90 @@ send_remote_request(Node, {IP, Port, Secret}, Request, Options) when ?GOOD_CMD(R
...
         {bad_pdu, "Authenticator Attribute is invalid" = Reason} ->
             update_client_response(bad_authenticator, MetricsInfo, Request),
             ?LOG(error, "~s INF: Noreply for request ~p. Could not decode the request, reason: ~s", [printable_peer(ServerIP, Port), Request, Reason]),
-            noreply;
+            maybe_failover(Request, noreply, {ServerIP, Port}, Options);
```
Comment (RoadRunnr) [inline src/eradius_client.erl:151]: this is also IMHO not a fail-over reason, a misconfigured shared secret is a local problem and should cause an immediate abort of the request

File: src/eradius_client.erl [hunk 25]
```diff
@@ -323,6 +382,31 @@ configure(State) ->
...
+    prepare_pool(Servers);
+prepare_pool([{Addr, Port, _} | Servers]) ->
+    store_radius_server_from_pool(Addr, Port, ?DEFAULT_RETRIES),
+    prepare_pool(Servers).
+
+%% private
+store_radius_server_from_pool(Addr, Port, Retries) when is_tuple(Addr) ->
+    ets:insert(?MODULE, {{Addr, Port}, Retries, Retries});
+store_radius_server_from_pool(Addr, Port, Retries) ->
+    case inet:gethostbyname(Addr) of
```
Comment (RoadRunnr) [inline src/eradius_client.erl:402]: configuring RADIUS peers by name is usually a problem. That originates from RFC 2865, Sect. 3 that states:

> A RADIUS server MUST use the source IP address of the RADIUS UDP packet to decide which shared secret to use

Name based configuration hide the IP and might therefore lead to hard find misconfigurations of the shared secret.

**Review by RoadRunnr — COMMENTED** [review]
I'm not too thrilled by the overall "feel" of this PR, but that goes for most of the eradius code in its current state. So adding another piece of code that I'm not feeling good about is probably not a problem.

The way the server pools are stored and handled feels wrong for lack of a better description. I would probably opted for per server process that could also do active monitoring (like FreeRADIUS does with Status-Server requests)

I don't see any big overall problems, so no reject, but also no accept from me. Feel free to ignore this comment altogether.

File: src/eradius_client.erl (hunk 24 above)
Comment (0xAX) [reply src/eradius_client.erl]: fixed

File: src/eradius_client.erl (hunk 23 above)
Comment (0xAX) [reply src/eradius_client.erl]: fixed

File: src/eradius_client.erl (hunk 25 above)
Comment (0xAX) [reply src/eradius_client.erl]: reworked this, it will throw error if not an IP address will be given.

**Comment by 0xAX** [comment]
> I'm not too thrilled by the overall "feel" of this PR, but that goes for most of the eradius code in its current state. So adding another piece of code that I'm not feeling good about is probably not a problem.
> 
> The way the server pools are stored and handled feels wrong for lack of a better description. I would probably opted for per server process that could also do active monitoring (like FreeRADIUS does with Status-Server requests)
> 
> I don't see any big overall problems, so no reject, but also no accept from me. Feel free to ignore this comment altogether.

@mgumz stated that not all upstream servers may provide support for Status-Server requests, thus it was done as it is now - ets based.
